File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    public int solution(int[][] maps) {
        int answer = -1;
        
        Queue<int[]> queue = new LinkedList<>();
        int row = maps.length;
        int col = maps[0].length;
        boolean[][] visited = new boolean[row][col];
        
        queue.offer(new int[] {0, 0, 1}); // x좌표, y좌표, 이동한 칸의 수
        visited[0][0] = true;
        
        int[] dx = new int[] {-1, 1, 0, 0};
        int[] dy = new int[] {0, 0, -1, 1};
        
        while (!queue.isEmpty()) {
            int[] arr = queue.poll();
            int x = arr[0];
            int y = arr[1];
            int count = arr[2];
            
            if (x == row - 1 && y == col - 1) {
                answer = count;
                break;
            }
            
            for (int i = 0; i < 4; i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];
                
                if (0 <= nx && nx < row && 0 <= ny && ny < col && maps[nx][ny] == 1 && !visited[nx][ny]) {
                    visited[nx][ny] = true;
                    queue.offer(new int[] {nx, ny, count + 1});
                }
            }
        }
        
        return answer;
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    
    row = len(maps)
    col = len(maps[0])
    queue = deque()
    queue.append((0, 0, 1)) # x좌표, y좌표, 칸 개수
    
    visited = [[False] * col for _ in range(row)]
    visited[0][0] = True
    
    dx = [-1, 1, 0, 0]
    dy = [0, 0, -1, 1]

    answer = -1
    while queue:
        x, y, count = queue.popleft()
        
        # 상대팀 진영에 도착
        if x == row - 1 and y == col - 1:
            answer = count
            break
        
        for i in range(4):
            nx, ny = x + dx[i], y + dy[i]
            
            if 0 <= nx < row and 0 <= ny < col and maps[nx][ny] == 1 and not visited[nx][ny]:
                visited[nx][ny] = True
                queue.append((nx, ny, count + 1))
        
    return answer
        